traitsui/qt4/video_editor.py

```diff
@@ -0,0 +1,362 @@
+        parent : QWidget or None
+            The parent widget for this widget.
+        """
+        self.control = QVideoWidget()
+        self.control.setSizePolicy(
+            QSizePolicy.Expanding, QSizePolicy.Expanding
+        )
+        self.control.setBackgroundRole(QPalette.Window)
+
+        self.media_player = QMediaPlayer(None, QMediaPlayer.VideoSurface)
+        self._set_video_url()
+        self.media_player.setVideoOutput(self.control)
+        self.media_player.setMuted(self.muted)
+        self._state_changed()
+        self._aspect_ratio_changed()
+        self._muted_changed()
+        self._volume_changed()
+        self._playback_rate_changed()
+
+        self._connect_signals()
+
+        self.set_tooltip()
+
+    def dispose(self):
+        self._disconnect_signals()
+
+    def update_editor(self):
+        """Update the editor when the object trait changes externally."""
+        self._set_video_url()
+
+    # ------------------------------------------------------------------------
+    # Private interface
+    # ------------------------------------------------------------------------
+
+    def _connect_signals(self):
+        if self.media_player is not None:
+            self.media_player.stateChanged.connect(self._state_changed_emitted)
+            self.media_player.positionChanged.connect(
+                self._position_changed_emitted)
+            self.media_player.durationChanged.connect(
+                self._duration_changed_emitted)
+            self.media_player.error.connect(self._error_emitted)
+            self.media_player.mediaStatusChanged.connect(
+                self._media_status_changed_emitted)
+            self.media_player.bufferStatusChanged.connect(
+                self._buffer_status_changed_emitted)
+
+    def _disconnect_signals(self):
+        if self.media_player is not None:
+            self.media_player.stateChanged.disconnect(
+                self._state_changed_emitted)
+            self.media_player.positionChanged.disconnect(
+                self._position_changed_emitted)
+            self.media_player.durationChanged.disconnect(
+                self._duration_changed_emitted)
+            self.media_player.error.disconnect(self._error_emitted)
+            self.media_player.mediaStatusChanged.disconnect(
+                self._media_status_changed_emitted)
+            self.media_player.bufferStatusChanged.disconnect(
+                self._buffer_status_changed_emitted)
+
+    def _set_video_url(self):
+        qurl = QUrl.fromUserInput(self.value)
+        if qurl.isValid():
+            self.media_content = QMediaContent(qurl)
+            self.control.updateGeometry()
+        else:
+            self.media_content = QMediaContent(None)
+            self.control.updateGeometry()
+
+    # Signal handlers -------------------------------------------------------
+
+    def _state_changed_emitted(self, state):
+        self.state = reversed_state_map[state]
```

Comment:
Should this be wrapped by a `self.updating_value()` context?

Reply:
Possibly - again, don't have the context loaded into my head, so it you think it is needed, do it!

Reply:
I'm leaving this alone for now. It's not as straightforward as `position` was.